packages/firestore/src/remote/persistent_stream.ts

```diff
@@ -157,16 +169,18 @@ export abstract class PersistentStream<
   private inactivityTimerPromise: CancelablePromise<void> | null = null;
   private stream: Stream<SendType, ReceiveType> | null = null;
 
-  protected backoff: ExponentialBackoff;
+  /** A sentinel object used to ignore auth callbacks for prior attempts. */
+  private currentAuthAttempt: object = null;
```

Comment:
I've addressed (1) elsewhere. I think what I was really asking for was a stream generation number, but read on.

Re (2) I think you're right: this is a mechanism for implementing centralized cancellation. We could instead force everything to be cancellation aware. We've partially done this on iOS by way of `FSTCallbackFilter`.

`dispatchIfStillActive` was always kind of a hack. It works because Javascript is so functional, but it's fundamentally a non-portable pattern. This came about back when we were seriously under the gun for time and I didn't push back on Jonny to stop and think about how to express this portably.

Functionally I think we have a few concerns to address in a centralized way:

  * we want to dispatch back onto the async queue
  * we need to be able to cancel the response if we don't need it anymore
  * cancelation needs to be checked as we're executing on the async queue

This is basically what `DelayedOperation` implements, but it does so in a non-reusable way. If we could share that between DelayedOperations and the auth callback we'd be golden on that front.

The stream callbacks could be treated the same way with something like `FSTCallbackFilter` here.